### PR TITLE
IGNITE-23615 Use shared threadpools in RocksDbFlusher for RocksDbKeyValueStorage

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-23615

### CODE REVIEW

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    private static String threadPrefix(InjectExecutorService injectExecutorService, String fieldName, Class<?> fieldType) {
+        String threadPrefix = injectExecutorService.threadPrefix();
+
+        if (threadPrefix != null && !"".equals(threadPrefix)) {
+            return threadPrefix;
+        }
+
+        return String.format("test-%s-%s", fieldType.getSimpleName(), fieldName);
```
Comment (rpuch) [inline modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java:227]: What if there are two fields named identically in a superclass and subclass? It seems their executors will get the same prefix

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java [hunk 2]
```diff
@@ -299,16 +301,12 @@ public RocksDbKeyValueStorage(
         );
 
         this.dbPath = dbPath;
+        this.scheduledExecutor = scheduledExecutor;
```
Comment (rpuch) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:304]: Is this executor used for scheduling absolutely any tasks, or for some specific tasks (like handling RocksDB-related stuff)? If the latter is true, would it make sense to rename it to reflect its purpose?

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java [hunk 3]
```diff
@@ -691,7 +691,8 @@ public class IgniteImpl implements Ignite {
                 name,
                 metastorageWorkDir.dbPath(),
                 failureManager,
-                readOperationForCompactionTracker
+                readOperationForCompactionTracker,
+                threadPoolsManager.commonScheduler()
```
Comment (rpuch) [inline modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java:695]: `threadPoolsManager.commonScheduler()` should only be used for light, fast tasks, never doing I/O, but here it's used for I/O, as it seems. Let's use another pool here

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+        @InjectExecutorService(
+                threadCount = 3,
+                threadPrefix = "test-bar-static-executor",
+                allowedOperations = {STORAGE_READ, STORAGE_WRITE}
+        )
+        private static ScheduledExecutorService staticScheduledExecutorService;
+
+        @InjectExecutorService
+        private ExecutorService fieldExecutorServiceWithDefaults;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java:70]: Renamed it a little differently.

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+        @InjectExecutorService(threadCount = 4, threadPrefix = "test-foo-field-executor", allowedOperations = STORAGE_WRITE)
+        private ExecutorService fieldExecutorService;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java:73]: Renamed it a little differently.

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+        @InjectExecutorService
+        private ScheduledExecutorService fieldScheduledExecutorServiceWithDefaults;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java:76]: Renamed it a little differently.

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+        @InjectExecutorService(threadCount = 5, threadPrefix = "test-bar-field-executor", allowedOperations = PROCESS_RAFT_REQ)
+        private ScheduledExecutorService fieldScheduledExecutorService;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java:79]: Renamed it a little differently.

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+            checkScheduledExecutorService(fieldScheduledExecutorService, 5, "test-bar-field-executor", PROCESS_RAFT_REQ);
+            checkExecutorService(parameterExecutorService, 6, "test-foo-param-executor", WAIT);
+            checkScheduledExecutorService(parameterScheduledExecutorService, 7, "test-bar-param-executor", WAIT);
+        }
+    }
+
+    @ExtendWith(ExecutorServiceExtension.class)
+    static class ErrorFieldInjectionTest {
+        @InjectExecutorService
+        private static Integer staticExecutorService;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java:135]: Renamed it a little differently.

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+        @InjectExecutorService
+        private String fieldExecutorService;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java:138]: Renamed it a little differently.

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+        @Test
+        public void test(
+                @InjectExecutorService
+                Boolean parameterExecutorService
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java:143]: Renamed it a little differently.

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+    ) {
+        assertEquals(executor.getCorePoolSize(), expCorePoolSize);
+
+        assertThat(
+                runAsync(() -> {
+                    Thread thread = Thread.currentThread();
+
+                    assertThat(thread, instanceOf(IgniteThread.class));
+
+                    IgniteThread thread1 = (IgniteThread) thread;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/ExecutorServiceExtensionTest.java:199]: I'll leave it as is.

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    private static void shutdownExecutors(ExtensionContext context, boolean forStatic) throws Exception {
+        List<ExecutorService> removed = (List<ExecutorService>) context.getStore(NAMESPACE).remove(storeKey(forStatic));
+
+        if (removed == null || removed.isEmpty()) {
+            return;
+        }
+
+        Stream<AutoCloseable> autoCloseableStream = removed.stream()
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java:156]: I don't think so, we'll see in the future.

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+    @Override
+    public Object resolveParameter(
+            ParameterContext parameterContext,
+            ExtensionContext extensionContext
+    ) throws ParameterResolutionException {
+        if (!supportsParameter(parameterContext, extensionContext)) {
+            throw new ParameterResolutionException("Unknown parameter:" + parameterContext.getParameter());
+        }
+
+        List<ExecutorService> executorServices = getOrCreateExecutorServiceListInStore(extensionContext, false);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java:104]: Yeah, maybe I can fix it.

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+
+        return false;
+    }
+
+    private static ExecutorService createExecutorService(Field field) {
+        InjectExecutorService injectExecutorService = field.getAnnotation(InjectExecutorService.class);
+
+        assert injectExecutorService != null : field;
+
+        return createExecutorService(injectExecutorService, field.getName(), field.getType(), field.getDeclaringClass());
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java:194]: Yes, do you have any suggestions?

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java (hunk 1 above)
Comment (tkalkirill) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java:227]: You're right, I'll fix it a little.

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java (hunk 3 above)
Comment (tkalkirill) [reply modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java:695]: I think that it is necessary to fix the fact that IO operations take place at `RocksDbFlusher#scheduledPool`.

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java (hunk 2 above)
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:304]: I'll add a description.

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/ExecutorServiceExtension.java]: Well, I asked a question :) What is already in code makes some sense already. Or we could use the real test class to obtain a logger. Up to you to decide.

**Review by rpuch — APPROVED** [review]

**Pull request merged**